### PR TITLE
Avoid PPEC crash during dithering ops

### PR DESCRIPTION
**Pull request by bwdev01**
This is the most restrictive approach because it only applies during dithering.  The only identified (rare) field problems occurred during dithering and settling.  Generalizing the protection beyond dithering would take more work.

### CODE REVIEW

File: src/star.cpp [hunk 1]
```diff
@@ -1031,8 +1031,9 @@ bool GuideStar::AutoFind(const usImage& image, int extraEdgeAllowance, int searc
             // We're repeating the find, so we're vulnerable to hot pixels and creation of unwanted duplicates
             if (tmp.WasFound() && tmp.SNR >= minSNR)
             {
-                bool duplicate = std::find_if(foundStars.begin(), foundStars.end(), [&tmp](const GuideStar& other)
-                                              { return CloseToReference(tmp, other); }) != foundStars.end();
+                bool duplicate =
+                    std::find_if(foundStars.begin(), foundStars.end(),
+                                 [&tmp](const GuideStar& other) { return CloseToReference(tmp, other); }) != foundStars.end();
```
Comment (agalasso) [inline src/star.cpp:1036]: this change seems to be unrelated to the pr ... let's revert it and get star.cpp out of the pr

File: contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp [hunk 2]
```diff
@@ -274,7 +274,17 @@ double GaussianProcessGuider::result(double input, double SNR, double time_step,
...
-        deduceResult(time_step); // just pretend we would do dark guiding...
+        try
+        {
+            deduceResult(time_step); // just pretend we would do dark guiding...
+        }
+        catch (std::runtime_error err)
+        {
+            reset();
+            std::string outStr = "PPEC: Model reset after exception: " + std::string(err.what());
+            GPDebug->Log(outStr.c_str());
```
Comment (agalasso) [inline contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp:285]: Log takes a printf-style format string so this could be a little dangerous (I'm thinking of the case where outStr contains a `%` character somewhere)
Here's is a safer way that also gets rid of the unneeded temporary string construction:
```suggestion
            GPDebug->Log("PPEC: Model reset after exception: %s", err.what());
```

File: contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp [hunk 3]
```diff
@@ -274,7 +274,17 @@ double GaussianProcessGuider::result(double input, double SNR, double time_step,
         {
             dithering_active_ = false;
         }
-        deduceResult(time_step); // just pretend we would do dark guiding...
+        try
+        {
+            deduceResult(time_step); // just pretend we would do dark guiding...
+        }
+        catch (std::runtime_error err)
```
Comment (agalasso) [inline contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp:281]: minor: can think of the catch args as like method arguments, and it is a good practice to pass objects by const reference where we want reference semantics (pass by reference) and do not need copy semantics (pass by value):
```suggestion
        catch (const std::runtime_error& err)
```

File: contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp [hunk 4]
```diff
@@ -689,38 +699,53 @@ Eigen::MatrixXd GaussianProcessGuider::regularize_dataset(const Eigen::VectorXd&
...
+
+                    gear_error_sum = 0.0;
+                    variance_sum = 0.0;
+                    ++j;
+                }
+            }
+        }
+        catch (std::runtime_error err)
+        {
+            throw err;
```
[the remark this answers is not in the document]
Comment (bwdev01) [reply contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp:748]: It won't compile without it.  The error message is that at least one handler must be specified for the try block in the function.

File: src/star.cpp (hunk 1 above)
Comment (bwdev01) [reply src/star.cpp:1036]: This is the same old problem with clang and star.cpp.   This has reverted the clang change without me ever opening or touching that file.  Something is seriously messed up with this.

File: contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (agalasso) [reply contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp:748]: right, both the try and catch both should be removed

File: src/star.cpp (hunk 1 above)
Comment (agalasso) [reply src/star.cpp:1036]: we'll need to find out what is changing that file in your local environment, and also be sure not to commit the file -- you can run `git restore src/star.cpp` before you commit it if you notice that it has been changed unintentionally.
Now that the "bad" version has been committed into the branch you can get the old version back with
```
git restore -s origin/master -- src/star.cpp
```
(or use any other version you want in place of `origin/master`)
Then `git add` `git commit` and `git push` to get the branch fixed up on github (origin).

**Comment by bwdev01** [comment]
Ok, you’ve lost me here.  Are you proposing that the throw shouldn’t be enclosed in a try/catch block?  The throw I’m forcing is at line 719 in that source.  Even if C++ allows a throw outside of a try/catch block, what happens?  The goal here is to trigger an exception before the crash occurs and have that exception caught in a function two levels down in the call stack at a point where recovery is well-defined.

**Comment by bwdev01** [comment]
Regarding the log output for the throw message.  Doing it the way you suggested results in this:

 

19:34:33.706 09.355 26228 PPEC: Model reset after exception: 湉敤⁸癯牥爭湵椠⁮敲畧慬楲敺摟瑡獡瑥ﴀ﷽ý0

 

I started out trying to use the %s format specifier and chased my tail trying to get anything usable – that’s how I ended up with my first try.  If the %s format is expecting a wxString, we’re out of luck because the gp* functions don’t use wxWidgets.  Any other approaches come to mind?

**Comment by agalasso** [comment]
> Regarding the log output for the throw message. Doing it the way you suggested results in this

@bwdev01 GPDebug::Log is not working correctly for `%s` arguments -- a latent bug this PR is now triggering it as it is the first use of `%s` in GPDebug.    I'm looking into it now...

**Comment by agalasso** [comment]
@bwdev01 if #1264 looks ok to you, we can merge that to master so you can make use of it in this pr, something like this:
```
    std::ostringstream message;
    message << "PPEC: Model reset after exception: " << err.what();
    GPDebug->Write(message.str().c_str());
```

File: contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp [hunk 5]
```diff
@@ -274,7 +274,19 @@ double GaussianProcessGuider::result(double input, double SNR, double time_step,
         {
             dithering_active_ = false;
         }
-        deduceResult(time_step); // just pretend we would do dark guiding...
+        try
+        {
+            deduceResult(time_step); // just pretend we would do dark guiding...
+        }
+        catch (const std::runtime_error err)
```
Comment (agalasso) [inline contributions/MPI_IS_gaussian_process/src/gaussian_process_guider.cpp:281]: we can use pass-by-reference semantics here rather than pass-by-value:
```suggestion
        catch (const std::runtime_error& err)
```

**Review by agalasso — COMMENTED** [review]
looks good, one minor suggestion

**Review by agalasso — APPROVED** [review]
nice work
